FastSimulation/TrackingRecHitProducer/src/PixelTemplateSmearerBase.cc

```diff
@@ -0,0 +1,1001 @@
+        {
+            throw cms::Exception("FastSimulation/TrackingRecHitProducer") << "Histogram ("<<theXHistN<<","<<theYHistN<<") was not found for PixelTemplateSmearer. Check if the smearing template exists.";
+        }
+
+
+        const SimpleHistogramGenerator* xgen = xgenIt->second;
+        const SimpleHistogramGenerator* ygen = ygenIt->second;
+
+        thePositionX = xgen->generate(random);
+        thePositionY = ygen->generate(random);
+
+
+        if( isForward )
+        {
+            thePositionY *= sign;
+        }
+        thePositionZ = 0.0; // set at the centre of the active area
+
+        thePosition = Local3DPoint(
+            simHit.localPosition().x() + thePositionX, 
+            simHit.localPosition().y() + thePositionY, 
+            simHit.localPosition().z() + thePositionZ
+        );
+        retry++;
+        if (retry > 10) 
+        {
+            // If we tried to generate thePosition, and it's out of the bounds
+            // for 10 times, then take and return the simHit's location.
+            thePosition = Local3DPoint(
+                simHit.localPosition().x(), 
+                simHit.localPosition().y(), 
+                simHit.localPosition().z()
+            );
+            break;
+        }
+    } while(fabs(thePosition.x()) > boundX  || fabs(thePosition.y()) > boundY);
+    
+    
+    FastSingleTrackerRecHit recHit(
+        thePosition,
+        theError,
+        *detUnit,
+        fastTrackerRecHitType::siPixel
+    );
+    return recHit;
+}
+
+
+TrackingRecHitProductPtr PixelTemplateSmearerBase::
+processUnmergedHits(
+    std::vector<TrackingRecHitProduct::SimHitIdPair> & unmergedHits, 
+    TrackingRecHitProductPtr product,
+    const PixelGeomDetUnit * detUnit,
+    const double boundX, const double boundY,
+    RandomEngineAndDistribution const * random
+) const
+{
+    for (auto simHitIdPair: unmergedHits)
+    {
+        FastSingleTrackerRecHit recHit = smearHit( *simHitIdPair.second, detUnit, boundX, boundY, random );
+        product->addRecHit( recHit ,{simHitIdPair});
+    }
+    return product;
+}
+
+
+TrackingRecHitProductPtr PixelTemplateSmearerBase::
+processMergeGroups(
+    std::vector< MergeGroup* > & mergeGroups,
+    TrackingRecHitProductPtr product,
+    const PixelGeomDetUnit * detUnit,
+    const double boundX, const double boundY,
+    RandomEngineAndDistribution const * random
+) const
+{
+    for ( auto mg_it = mergeGroups.begin(); mg_it != mergeGroups.end(); ++mg_it) 
+    {
+        if ((*mg_it)->smearIt)
+        {
+            FastSingleTrackerRecHit recHit = smearMergeGroup( *mg_it, detUnit, boundX, boundY, random );
+            product->addRecHit( recHit, (*mg_it)->group);
+        }
+    }
+    return product;
+}
+
+
+FastSingleTrackerRecHit PixelTemplateSmearerBase::
+smearMergeGroup(
+    MergeGroup* mg,
+    const PixelGeomDetUnit * detUnit,
+    const double boundX, const double boundY,
+    RandomEngineAndDistribution const * random
+) const 
```

Comment:
this method copies some code from the `smearHit` method. Can it be moved into an extra method?